gprofiler/java.py

```diff
@@ -33,66 +34,214 @@
+
+        self._check_disk_requirements()
+
+        # make out & log paths writable for all, so target process can write to them.
+        # see comment on TemporaryDirectoryWithMode in GProfiler.__init__.
+        touch_path(self._output_path_host, self.OUTPUTS_MODE)
+        self._recreate_log()
+        # copy libasyncProfiler.so if needed
+        if not os.path.exists(self._libap_path_host):
+            self._copy_libap()
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # ignore_errors because we are deleting paths via /proc/pid/root - and the process
+        # might have gone down already.
+        # remove them as best effort.
+        remove_path(self._output_path_host, missing_ok=True)
+        remove_path(self._log_path_host, missing_ok=True)
+
+    def _copy_libap(self) -> None:
+        # copy *is* racy with respect to other processes running in the same namespace, because they all use
+        # the same directory for libasyncProfiler.so, as we don't want to create too many copies of it that
+        # will waste disk space.
+        # my attempt here is to produce a race-free way of ensuring libasyncProfiler.so was fully copied
+        # to a *single* path, per namespace.
+        # newer kernels (>3.15 for ext4) have the renameat2 syscall, with RENAME_NOREPLACE, which lets you
+        # atomically move a file without replacing if the target already exists.
+        # this function operates similarly on directories. if you rename(dir_a, dir_b) then dir_b is replaced
+        # with dir_a iff it's empty. this gives us the same semantics.
+        # so, we create a temporary directory on the same filesystem (so move is atomic) in a race-free way
+        # by including the PID in its name; then we move it.
+        # TODO: if we ever move away from the multithreaded model, we can get rid of this complexity
+        # by ensuring a known order of execution.
+        ap_dir_host_tmp = f"{self._ap_dir_host}.{self.process.pid}"
+        os.makedirs(ap_dir_host_tmp)
+        libap_tmp = os.path.join(ap_dir_host_tmp, "libasyncProfiler.so")
+        shutil.copy(resource_path("java/libasyncProfiler.so"), libap_tmp)
+        os.chmod(libap_tmp, 0o755)  # make it accessible for all; needed with PyInstaller, which extracts files as 0700
+        try:
+            os.rename(ap_dir_host_tmp, self._ap_dir_host)
+        except OSError as e:
+            if e.errno == errno.ENOTEMPTY:
+                # remove our copy
+                shutil.rmtree(ap_dir_host_tmp)
+                # it should have been created by somene else.
+                assert os.path.exists(self._libap_path_host)
+            else:
+                raise
+
+    def _recreate_log(self) -> None:
+        touch_path(self._log_path_host, self.OUTPUTS_MODE)
+
+    def _check_disk_requirements(self) -> None:
+        free_disk = psutil.disk_usage(self._ap_dir_host).free
+        if free_disk < 250 * 1024:
```

Comment:
Please explain in a comment the 250kb (I guess it's the ap itself + log + outputs) estimation

Reply:
Yeah something like that. Just a limit to avoid killing the disk

Reply:
> Please explain in a comment


?

Reply:
explained?